Modelica/Resources/C-Sources/ModelicaTime.c

```diff
@@ -0,0 +1,245 @@
+
+   2. Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+
+   3. Neither the name of the copyright holder nor the names of its
+      contributors may be used to endorse or promote products derived from
+      this software without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+   ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+   FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+   SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+   CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+   OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// TODO: Remove when ModelicaTime.c will become part of ModelicaExternalC
+#if defined(DYMOSIM)
+#pragma once
+#endif
+
+#include "ModelicaTime.h"
+#include "ModelicaUtilities.h"
+#include "strptime.h"
+#include "repl_str.h"
+
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <time.h>
+#if !defined(NO_LOCALE)
+#include <locale.h>
+#endif
+
+#if !defined(NO_TIME)
+time_t epoch(int sec, int min, int hour, int mday, int mon, int year) {
+    struct tm tlocal;
+    time_t calendarTime;
+
+    memset(&tlocal, 0, sizeof(struct tm));
+    tlocal.tm_sec = sec;
+    tlocal.tm_min = min;
+    tlocal.tm_hour = hour;
+    tlocal.tm_mday = mday;
+    tlocal.tm_mon = mon - 1;
+    tlocal.tm_year = year - 1900;
+
+    calendarTime = mktime(&tlocal);
+    if (-1 == calendarTime) {
+        ModelicaFormatError("Not possible to convert \"%4i-%02i-%02i %02i:%02i:%02i\" "
+            "to time_t type in local time zone.\n", year, mon, mday, hour, min, sec);
+    }
+    return calendarTime;
+}
+#endif
+
+double ModelicaTime_difftime(int ms, int sec, int min, int hour, int mday,
+                             int mon, int year, int refYear) {
+    /* Get elapsed seconds w.r.t. reference year */
+#if defined(NO_TIME)
+    return 0.0;
+#else
+    const time_t endTime = epoch(sec, min, hour, mday, mon, year);
+    if (1970 == refYear) {
+        /* Avoid calling mktime for New Year 1970 in local time zone */
+        const time_t startTime = epoch(0, 0, 0, 2, 1, 1970);
+        return difftime(endTime, startTime) + 86400.0 + ms/1000.0;
+    }
+    else {
+        const time_t startTime = epoch(0, 0, 0, 1, 1, refYear);
+        return difftime(endTime, startTime) + ms/1000.0;
+    }
+#endif
+}
```

Comment:
There seem to be some troubles. With Visual Sutdio 2015 in Dymola 2020x these example do not work:

    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 0, 1, 1, 1970);       // does not work
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 0, 1, 1, 1969, 1969)  // does not work
    
For later years it works
    
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 0, 1, 1, 1971, 1971); // = 0.0
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 0, 1, 1, 2019, 2019); // = 0.0

Adding 1h also fixes the problem.
    
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 1, 1, 1, 1970);  // = 3600.0

Negative diffTime also has some troubles:
    
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 1, 1, 1, 1969);       // does not work    
    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 0, 1, 1, 1970, 1971); // does not work

Again, adding 1h fixes the problem.

    Modelica.Utilities.Internal.Time.diffTime(0, 0, 0, 1, 1, 1, 1970, 1971); // = -31532400.0

Reply:
Many of the corner cases are addressed by 2bc5990c08f32e10fc7944879f348c44cf6594c1, escept that `mktime` does not succeed for years earlier than 1970.